haystack/preview/components/preprocessors/text_document_cleaner.py

```diff
@@ -0,0 +1,205 @@
+import re
+from copy import deepcopy
+from functools import partial, reduce
+from itertools import chain
+from typing import Any, Dict, Generator, List, Optional, Set
+
+from haystack.preview import Document, component, default_from_dict, default_to_dict
+
+
+@component
+class TextDocumentCleaner:
+    """
+    Makes text documents more readable by cleaning empty lines, extra whitespaces, headers and footers, etc.
+    This is useful for preparing the documents for further processing by LLMs.
+    """
+
+    def __init__(
+        self,
+        remove_empty_lines: bool = True,
+        remove_extra_whitespaces: bool = True,
+        remove_repeated_substrings: bool = False,
+        remove_substrings: Optional[List[str]] = None,
+        remove_regex: Optional[str] = None,
+    ):
+        """
+        :param remove_empty_lines: Whether to remove empty lines.
+        :param remove_extra_whitespaces: Whether to remove extra whitespaces.
+        :param remove_repeated_substrings: Whether to remove repeated substrings, such as headers and footers.
+        :param remove_substrings: List of substrings to remove from the text.
+        :param remove_regex: Regex to match and replace substrings by "".
+        """
+
+        self.remove_empty_lines = remove_empty_lines
+        self.remove_extra_whitespaces = remove_extra_whitespaces
+        self.remove_repeated_substrings = remove_repeated_substrings
+        self.remove_substrings = remove_substrings
+        self.remove_regex = remove_regex
+
+    @component.output_types(documents=List[Document])
+    def run(self, documents: List[Document]):
+        if not isinstance(documents, list) or documents and not isinstance(documents[0], Document):
+            raise TypeError("TextDocumentCleaner expects a List of Documents as input.")
+
+        cleaned_docs = []
+        for doc in documents:
+            if doc.text is None:
+                raise ValueError(
+                    f"TextDocumentCleaner only works with text documents but document.text for document ID {doc.id} is None."
+                )
+            text = doc.text
+
+            if self.remove_empty_lines:
+                text = self._remove_empty_lines(text)
+            if self.remove_extra_whitespaces:
+                text = self._remove_extra_whitespaces(text)
+            if self.remove_repeated_substrings:
+                text = self._remove_repeated_substrings(text)
+            if self.remove_substrings:
+                text = self._remove_substrings(text, self.remove_substrings)
+            if self.remove_regex:
+                text = self._remove_regex(text, self.remove_regex)
+
+            cleaned_docs.append(Document(text=text, metadata=deepcopy(doc.metadata)))
```

Comment:
How would the id be updated if we don't create a new Document? With the current Document implementation I can't really trigger the id to be updated after I change the text. I would keep it as is for this reason. Can be updated later if we change the implementation of document.

Reply:
I'm torn.

On the one hand, it is strange to change the ID of the document just because of cleaning it.
On the other, I understand that we want to keep the Id tied to the content.

(not blocking, however...)

Reply:
I would really keep the id tied to the content, yes. If there are use cases that need to track the source of the document, one could track that source document id through metadata. Similar to how one could track the source file id through metadata.